File: Leetcode/Python/_28.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) == 0 and len(haystack) == 0:
            return 0
        window = len(needle)
        for i in range(len(haystack) + 1 - window):
            if haystack[i:i + window] == needle:
                return i
        return -1
      
      
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if len(haystack) < len(needle):
            return -1

        mapping_needle = {}
        mapping = {}
        for char in needle:
            if char not in mapping_needle:
                mapping_needle[char] = 0
            mapping_needle[char] += 1

        for i in range(len(needle)):
            char = haystack[i]
            if char not in mapping:
                mapping[char] = 0
            mapping[char] += 1

        if mapping_needle == mapping and haystack[:len(needle)] == needle:
            return 0
        left = 0
        for right in range(len(needle), len(haystack)):
            remove_char = haystack[left]
            mapping[remove_char] -= 1
            if mapping[remove_char] == 0:
                del mapping[remove_char]

            add_char = haystack[right]
            if add_char not in mapping:
                mapping[add_char] = 0
            mapping[add_char] += 1
            left += 1

            if mapping == mapping_needle and haystack[left:right + 1] == needle:
                return left
        return -1

File: Leetcode/cpp/28.cpp
class Solution {
public:
    std::vector<int> buildLPS(const std::string& needle) {
        int n = needle.length();
        std::vector<int> prefix(n, 0);
        int index = 1;
        int length = 0;
        while (index < n) {
            if (needle[index] == needle[length]) {
                prefix[index++] = ++length;
            } else {
                if (length != 0) {
                    length = prefix[length - 1];
                } else {
                    prefix[index++] = 0;
                }
            }
        }
        return prefix;
    }

    int strStr(std::string haystack, std::string needle) {
        std::vector<int> lps = buildLPS(needle);
        int i = 0;
        int n = haystack.length();
        int m = needle.length();
        int j = 0;
        while (i < n) {
            if (haystack[i] == needle[j]) {
                ++i;
                ++j;

                if (j == m) {
                    return i - j;
                }
            } else {
                if (j != 0) {
                    j = lps[j - 1];
                } else {
                    ++i;
                }
            }
        }
        return -1;
    }
};

class SolutionNotOptimal {
public:
    int strStr(string haystack, string needle) {
        int m = needle.length();
        int n = haystack.length();

        for (int windowStart = 0; windowStart <= n - m; windowStart++) {
            for (int i = 0; i < m; i++) {
                if (needle[i] != haystack[windowStart + i]) {
                    break;
                }
                if (i == m - 1) {
                    return windowStart;
                }
            }
        }

        return -1;
    }
};
